### DISCUSSION TITLE
add custom claim/attribute to the (access) token response

### DISCUSSION THREAD

**Discussion by XcrigX** [keycloak/keycloak#13239] [Q&A]
I have a use-case where I need to add a custom claim/attribute in the access token response itself: 
i.e. : 
{
    "access_token": "jwt-access-token",
    "id_token": "jwt-id-token",
    "refresh_token": "opaque-refresh-token",
    "customField": "custom-value"
}

The custom-value in this case would come from a user attribute. 

I see that the 'User Session Note' mapper has a checkbox for 'Add to access token response', but none of the other mapper types have this option. 
I can't find a way to set a user-session note without writing a custom authenticator, and I'd prefer to avoid that if possible.  

So my questions are:
a) is there a way to do this out of the box -  a way to set a user-session note from a user attribute or via another mapper - or some other way to get a custom attribute added

b) why does 'User Session Note' mapper have this feature but none of the other mappers do ?

Thanks.

**Comment by XcrigX** [comment]
FYI - This was the pull request that added the feature to the user session note mapper: 
https://github.com/keycloak/keycloak/pull/7371
So that mostly answers (B).

**Comment by sventorben** [comment]
Have you tried using a `User Attribute Mapper`? This should work ootb.

**Reply by XcrigX** [reply]
"User Attribute Mapper" will let you set a value in the access_token and id_token, but not set a value in the access token response itself. 
I want to set an additional sibling attribute to the access_token, id_token, etc which are included in the token response.

**Comment by eanea** [comment]
@XcrigX 
did you find a solution?

**Reply by XcrigX** [reply]
I don't think I did. I think I used a custom authenticator which set a user session note, and then a user session note mapper to add the user session note to the response.

**Comment by kcemanes** [comment]
> I don't think I did. I think I used a custom authenticator which set a user session note, and then a user session note mapper to add the user session note to the response.

@XcrigX I'm going through the same pains with a basic SMART on FHIR integration I'm doing right now and was wondering if you could help point me to what this might look like e.g. a custom Keycloak authentication extension(?). I'm guessing it'll look a lot like Alvearie's (https://github.com/Alvearie/keycloak-extensions-for-fhir) implementation?

I am also reluctant in creating custom code to do this. I'm not against starting from scratch but a bit of pointing to the right direction is always appreciated. Thanks.

**Reply by XcrigX** [reply]
Yes, it will look like Alvearie's solution.  At some level you'll likely need customization regardless to correlate a user identity to a FHIR Patient/User - so I don't know that you can avoid writing some custom code. 
Also, in case you are unaware, there is some discussion here on Keycloak for SMART that may give you some more context:
https://chat.fhir.org/#narrow/stream/179170-smart/topic/Keycloak.20for.20SMART.20authz
